File: packages/discord/src/commands/subscribe.ts
import { PREFIX } from "@daily-news-discord/environment";
import { Command } from "../utils/Command";
import { categoryArray } from "@daily-news-discord/news";
import {
  createGuildChannel,
  createGuildChannels,
  deleteGuildChannels,
  getGuildChannel,
} from "@daily-news-discord/database";

export default new Command({
  name: "subscribe",
  description: "Subscribe News to this channel",
  aliases: ["sub"],
  usage: `${PREFIX}subscribe <category>`,
  execute: async (msg, args) => {
    if (!msg.guild) {
      msg.reply("Please use this command in a server");
      return;
    }

    const category = args[0];
    if (!category) {
      msg.reply("Please specify a category");
      return;
    }

    // Check if category is valid
    if (!categoryArray.includes(category) && category !== "all") {
      msg.reply("Please specify a valid category");
      return;
    }

    const channel = msg.channel;
    const guild = msg.guild;

    if (category === "all") {
      await deleteGuildChannels(guild.id, channel.id);
      await createGuildChannels(guild.id, channel.id, categoryArray);
      msg.reply("You are now subscribed to all categories");
      return;
    }

    const guildChannel = await getGuildChannel(guild.id, channel.id, category);
    if (guildChannel) {
      msg.reply("You are already subscribed to this channel");
      return;
    }

    await createGuildChannel(guild.id, channel.id, category);
    msg.reply(`You are now subscribed to ${category}`);
  },
});

File: commands/subscribe.go
package commands

import (
	"log"
	"strings"

	"github.com/SushiWaUmai/daily-news-discord/db"
	"github.com/SushiWaUmai/daily-news-discord/news"
	"github.com/bwmarrin/discordgo"
)

func init() {
	var categoryChoices []*discordgo.ApplicationCommandOptionChoice
	for _, c := range news.Categories {
		categoryChoices = append(categoryChoices, &discordgo.ApplicationCommandOptionChoice{
			Name:  c,
			Value: c,
		})
	}
	categoryChoices = append(categoryChoices, &discordgo.ApplicationCommandOptionChoice{
		Name:  "all",
		Value: "all",
	})

	createCommand(&discordgo.ApplicationCommand{
		Name:        "subscribe",
		Description: "subscribes news to this channel",
		Options: []*discordgo.ApplicationCommandOption{
			{
				Type:        discordgo.ApplicationCommandOptionString,
				Name:        "category",
				Description: "The news category you want to subscribe to",
				Required:    true,
				Choices:     categoryChoices,
			},
		},
	}, func(dg *discordgo.Session, i *discordgo.InteractionCreate) {
		guildID := i.GuildID
		channelID := i.ChannelID
		category := strings.ToLower(i.ApplicationCommandData().Options[0].StringValue())

		if category == "all" {
			db.DeleteGuildChannels(guildID, channelID)
			db.CreateGuildChannels(guildID, channelID, news.Categories)

			dg.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
				Type: discordgo.InteractionResponseChannelMessageWithSource,
				Data: &discordgo.InteractionResponseData{
					Content: "You are now subscribed to all categories",
				},
			})
			return
		}

		guildChannel, _ := db.GetGuildChannel(guildID, channelID, category)
		if guildChannel != nil {
			dg.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
				Type: discordgo.InteractionResponseChannelMessageWithSource,
				Data: &discordgo.InteractionResponseData{
					Content: "You are already subscribed to this category",
				},
			})
			return
		}

		err := db.CreateGuildChannel(guildID, channelID, category)
		if err != nil {
			log.Println("Failed to create GuildChannel", err)
			return
		}

		dg.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
			Type: discordgo.InteractionResponseChannelMessageWithSource,
			Data: &discordgo.InteractionResponseData{
				Content: "You are now subscribed to " + category,
			},
		})
	})
}
